File: Natural Language Processing.py
# Natural Language Processing

# Import the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
ds = pd.read_csv("Restaurant_Reviews.tsv", delimiter = '\t', quoting = 3) # Quoting is for ignoring double quotes

# Cleaning the text
import re
import nltk
# nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

corpus = []
for i in range(0, 1000):
    review = re.sub('[^a-zA-Z]', ' ', ds['Review'][i]) # Keep only letters
    review = review.lower() # To lowercase
    review = review.split() # Split into an array of individual words
    # Remove usless words
    ps = PorterStemmer()
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    review = ' '.join(review,) # Join the words back together into a string
    corpus.append(review)
    
    
# Creating the bag of words model
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500) # Taking the top 1500 words
X = cv.fit_transform(corpus).toarray()
Y = ds.iloc[:, 1].values

####### Naive Bayes

# Splitting the dataset into a training set and test set
from sklearn.cross_validation import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.20, random_state = 0)


# Fitting Naive Bayes to the Training Set
from sklearn.naive_bayes import GaussianNB
cl = GaussianNB()
cl.fit(X_train, Y_train)

# Predicint the test set results
y_pred = cl.predict(X_test)

# Making the confustion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, y_pred)


####### End Naive Bayes

####### Decision Tree Classification

# Splitting the dataset into a training set and test set
from sklearn.cross_validation import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.20, random_state = 0)


# Fitting Decision Tree Classification to the Training Set
from sklearn.tree import DecisionTreeClassifier
cl = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
cl.fit(X_train, Y_train)

# Predicint the test set results
y_pred = cl.predict(X_test)

# Making the confustion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, y_pred)

####### End Decision Tree Classification

####### Random Forest Classification
# Splitting the dataset into a training set and test set
from sklearn.cross_validation import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.20, random_state = 0)


# Fitting the Random Forest classifier to the Training Set
from sklearn.ensemble import RandomForestClassifier
cl = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
cl.fit(X_train, Y_train)

# Predicint the test set results
y_pred = cl.predict(X_test)

# Making the confustion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, y_pred)

####### End Random Forest Classification

File: Natural Language Processing.R
# Natural Language Processing

# Importing the dataset
ds = read.delim('Restaurant_Reviews.tsv', quote = '', stringsAsFactors = FALSE)

# Cleaning the text
# install.packages('tm')
# install.packages('SnowballC')
library(tm)
library(SnowballC)
corpus = VCorpus(VectorSource(ds$Review))
corpus = tm_map(corpus, content_transformer(tolower)) # toLowercase
corpus = tm_map(corpus, removeNumbers)
corpus = tm_map(corpus, removePunctuation)
corpus = tm_map(corpus, removeWords, stopwords()) # Remove unnessisary words
corpus = tm_map(corpus, stemDocument)
corpus = tm_map(corpus, stripWhitespace)

# Creating a bag word model
dtm = DocumentTermMatrix(corpus)
dtm = removeSparseTerms(dtm, 0.999) # Keeps the top 99.9% of words.
ds_class = as.data.frame(as.matrix(dtm))
ds_class$Liked = ds$Liked

##### Using the Random Forest Classification #####

# Encoding the target feature as factor
ds_class$Liked = factor(ds_class$Liked, levels = c(0, 1))

# Splitting the dataset into a training set and test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(ds_class$Liked, SplitRatio = 0.8)
trs = subset(ds_class, split == TRUE) # Training Set
tes = subset(ds_class, split == FALSE) # Test Set

# Fitting the Random Forest Classification to the training set
# install.packages('randomForest')
library(randomForest)
cl = randomForest(x = trs[-692],
                  y = trs$Liked,
                  ntree = 10)

# Predicting the Test set Results
y_pred = predict(cl, newdata = tes[-692])

# Making the Confusion Matrix
cm = table(tes[, 692], y_pred)
